src/peft/tuners/lora.py

```diff
@@ -1193,8 +1204,48 @@ def __init__(
             self.update_layer(adapter_name, r, lora_alpha, lora_dropout, init_lora_weights)
             self.active_adapter = adapter_name
 
+        def merge(self):
+            if self.active_adapter not in self.lora_A.keys():
+                return
+            if self.merged:
+                warnings.warn("Already merged. Nothing to do.")
+                return
+            if self.r[self.active_adapter] > 0:
+                warnings.warn(
+                    "Merge lora module to 4-bit linear may get different generations due to rounding errors."
+                )
+                kwargs = self.weight.__dict__
+                lora_data = self.get_delta_weight(self.active_adapter)
+                w_data = bnb.functional.dequantize_4bit(self.weight.data, self.weight.quant_state) + lora_data
```

Comment:
I am not sure but I think that you need to specify the `quant_type` here and below
```suggestion
                w_data = bnb.functional.dequantize_4bit(self.weight.data, self.weight.quant_state, quant_type=self.weight.quant_type) + lora_data
```

Reply:
Thanks for your comment. Actually, there is no need to pass `quant_type` because it is already in [quant_state](https://github.com/TimDettmers/bitsandbytes/blob/main/bitsandbytes/functional.py#L898).

I have added a comment that refer to [https://gist.github.com/ChrisHayduk/1a53463331f52dca205e55982baf9930](https://gist.github.com/ChrisHayduk/1a53463331f52dca205e55982baf9930) before 4bit merge

Reply:
Thanks a lot!